pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -0,0 +1,325 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/jsonpb"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"fmt"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	adsDebug = os.Getenv("PILOT_DEBUG_ADS") != "0"
+
+	adsClientsMutex sync.RWMutex
+	adsClients      = map[string]*XdsConnection{}
+)
+
+// XdsConnection is a listener connection type.
+type XdsConnection struct {
+	// PeerAddr is the address of the client envoy, from network layer
+	PeerAddr string
+
+	// Time of connection, for debugging
+	Connect time.Time
+
+	// Node is the name of the remote node
+	Node string
+
+	modelNode *model.Proxy
+
+	// Sending on this channel results in  push. We may also make it a channel of objects so
+	// same info can be sent to all clients, without recomputing.
+	pushChannel chan struct{}
+
+	// TODO: migrate other fields as needed from model.Proxy and replace it
+
+	//HttpConnectionManagers map[string]*http_conn.HttpConnectionManager
+
+	HTTPListeners []*xdsapi.Listener
```

Comment:
why just HTTP listeners?

Reply:
didn't have time to add the others yet...